enzyme/Enzyme/FunctionUtils.cpp

```diff
@@ -1632,8 +1617,31 @@ Function *PreProcessCache::CloneFunctionWithReturns(
     RetType = Type::getVoidTy(RetType->getContext());
 
   // Create a new function type...
-  FunctionType *FTy =
-      FunctionType::get(RetType, ArgTypes, F->getFunctionType()->isVarArg());
+  return FunctionType::get(RetType, ArgTypes, FTy->isVarArg());
+}
+
+Function *PreProcessCache::CloneFunctionWithReturns(
+    DerivativeMode mode, Function *&F, ValueToValueMapTy &ptrInputs,
+    const std::vector<DIFFE_TYPE> &constant_args,
+    SmallPtrSetImpl<Value *> &constants, SmallPtrSetImpl<Value *> &nonconstant,
+    SmallPtrSetImpl<Value *> &returnvals, ReturnType returnValue, Twine name,
+    ValueToValueMapTy *VMapO, bool diffeReturnArg, llvm::Type *additionalArg) {
+  assert(!F->empty());
+  F = preprocessForClone(F, mode);
+  llvm::ValueToValueMapTy VMap;
+  llvm::FunctionType *FTy =
+      getFunctionTypeForClone(F->getFunctionType(), additionalArg,
+                              constant_args, diffeReturnArg, returnValue);
+
+  for (BasicBlock &BB : *F) {
+    for (Instruction &I : BB) {
```

Comment:
For efficiency sake, can you just use the terminator of the block, rather than looping through all instructions?

Reply:
done